src/ripple/app/misc/impl/TxQ.cpp

```diff
@@ -745,189 +891,226 @@ TxQ::apply(Application& app, OpenView& view,
-
-                        if (feeLevelPaid <= requiredMultiLevel)
-                        {
-                            // Drop the current transaction
-                            JLOG(j_.trace()) <<
-                                "Ignoring transaction " <<
-                                transactionID <<
-                                ". Needs fee level of " <<
-
-                                requiredMultiLevel <<
-                                ". Only paid " <<
-                                feeLevelPaid;
-                            return{ telINSUF_FEE_P, false };
-                        }
-                    }
-                    if (workingIter->first == tSeq)
-                    {
-                        // If we're replacing this transaction, don't
-                        // count it.
-                        assert(replacedItemDeleteIter);
-                        multiTxn->includeCurrentFee =
-                            std::next(workingIter) !=
-                                txQAcct.transactions.end();
-                        continue;
-                    }
-                    // Don't worry about the blocker status of txs
-                    // after the current.
-                    if (workingIter->first < tSeq &&
-                        workingIter->second.consequences().isBlocker())
-                    {
-                        // Drop the current transaction, because it's
-                        // blocked by workingIter.
-                        JLOG(j_.trace()) <<
-                            "Ignoring transaction " <<
-                            transactionID <<
-                            ". A blocker-type transaction " <<
-                            "is in the queue.";
-                        return{ telCAN_NOT_QUEUE_BLOCKED, false };
-                    }
-                    multiTxn->fee +=
-                        workingIter->second.consequences().fee();
-                    multiTxn->potentialSpend +=
-                        workingIter->second.consequences().potentialSpend();
+                    if (txSeqOrT < acctSeqOrT)
+                        return {tefPAST_SEQ, false};
+                    else if (txSeqOrT > acctSeqOrT)
+                        return {terPRE_SEQ, false};
                 }
-                if (workingSeq < tSeq)
-                    // Transactions are missing before `tx`.
-                    multiTxn.reset();
+            }
+            else if (! replacedTxIter)
+            {
+                // The current transaction is not replacing a transaction
+                // in the queue.  So apparently there's a transaction in
+                // front of this one in the queue.  Make sure the two
+                // transactions interact well.
+                TER const ter {
+                    prevIter->second.consequences().couldBeNext (txSeqOrT)};
+                if (!isTesSuccess (ter))
+                    return {ter, false};
```

Comment:
Nice!  This change helped me find a bug in `nextQueuableSeq()`.  As is noted elsewhere in `apply` sometimes transactions are present both in the queue and in the ledger.  When that happens `nextQueuableSeq()` would do the wrong thing.  So I've recoded `nextQueuableSeq()` like this:
```
// The passed in proposedSeq is typically (but not required to be) the
// sequence of the account root of account.  The goal is to return a
// SeqProxy for a sequence that will fill the next available hole
// in the queue for the passed in account.
//
// Typically there will be no queued transactions.  In that case the
// proposedSeq is returned as a SeqProxy.
//
// If there are queued transactions for the account then the first viable
// sequence number, that is not used by a transaction in the queue, must
// be found and returned.
SeqProxy
TxQ::nextQueuableSeqImpl (
    ReadView const& view,
    AccountID const& account,
    std::uint32_t proposedSeq,
    std::lock_guard<std::mutex> const&) const
{
    SeqProxy const proposed = SeqProxy::sequence (proposedSeq);

    // If the account is not in the queue, the proposed start is good enough.
    auto const accountIter = byAccount_.find(account);
    if (accountIter == byAccount_.end() ||
            accountIter->second.transactions.empty())
        return proposed;

    // If the account is not in the ledger the proposed start is good enough.
    auto const sleAccount = view.read (keylet::account (account));
    if (! sleAccount)
        return proposed;

    TxQAccount::TxMap const& acctTxs = accountIter->second.transactions;

    // Ignore any sequence-based queued transactions that slipped into the
    // ledger while we were not watching.  This does actually happen in the
    // wild, but it's uncommon.
    SeqProxy const acctSeqProx =
        SeqProxy::sequence ((*sleAccount)[sfSequence]);

    TxQAccount::TxMap::const_iterator txIter =
        acctTxs.lower_bound (acctSeqProx);

    if (txIter == acctTxs.end() ||
            !txIter->first.isSeq() ||
            txIter->first != acctSeqProx)
        // Either there are no queued sequence-based transactions equal to or
        // following acctSeqProx or acctSeqProx is not currently in the queue.
        // So acctSeqProx is as good as it gets.
        return acctSeqProx;

    // There are sequence-based transactions queued that follow acctSeqProx.
    // Locate the first opening to put a transaction into.
    SeqProxy attempt = txIter->second.consequences().followingSeq();
    while (++txIter != acctTxs.cend())
    {
        if (attempt < txIter->first)
            break;

        attempt = txIter->second.consequences().followingSeq();
    }
    return attempt;
}
```
Let me know if you spot any problems in that.  Thanks!

Reply:
The only thing that jumps out at me is the same idea I put out in another comment instead of looking it up in the middle of the function, pass in `SLE const& sleAccount`, and get the information `AccountID` and sequence out of it. I know the comment says the proposed seq is not required to be the account sequence, but why allow that, when all it does is open up more potential to give the wrong answer? I'd rather know that the function is more likely to give me what it promises to give me: the next queuable sequence.